lms/services/lti_h.py

```diff
@@ -44,71 +28,64 @@ def __init__(self, _context, request):
         """
-        Create or update the Hypothesis user for the request's LTI user.
+        Sync standard data to H for an LTI launch with the provided groups.
 
-        Update the h user's information from LTI data. If the user doesn't
-        exist yet, call the h API to create one.
+        This will upsert the provided list of groups, the current user and
+        make that user a member of each group.
+
+        :param groups: A list of Group objects.
         """
 
+        if not self._context.provisioning_enabled:  # pylint: disable=protected-access
+            return None
+
+        try:
+            return self._sync_to_h(groups)
+
+        except HAPIError as err:
+            raise HTTPInternalServerError(explanation=err.explanation) from err
+
+    def _sync_to_h(self, groups):
+        h_user = self._context.h_user
+
         self.h_api.upsert_user(
-            h_user=self._context.h_user,
+            h_user=h_user,
             provider=self._context.h_provider,
             provider_unique_id=self._context.h_provider_unique_id,
         )
 
-    @lti_h_action
-    def upsert_course_group(self):
-        """
-        Create or update the Hypothesis group for the request's LTI course.
-
-        Call the h API to create a group for the LTI course, if one doesn't
-        exist already.
+        for group in groups:
+            self._upsert_h_group(group, creator=self._context.h_user)
 
-        Groups can only be created if the LTI user is allowed to create
-        Hypothesis groups (for example instructors are allowed to create
-        groups).
-
-        If the group for the course hasn't been created yet, and if the user
-        isn't allowed to create groups (e.g. if they're just a student) then
-        show an error page instead of continuing with the LTI launch.
-        """
+            self.group_info_service.upsert(
+                authority_provided_id=self._context.h_authority_provided_id,
+                consumer_key=self._request.lti_user.oauth_consumer_key,
+                params=self._request.params,
+            )
 
-        self._upsert_h_group(
-            group_id=self._context.h_groupid,
-            group_name=self._context.h_group_name,
-            creator=self._context.h_user,
-        )
-
-        self.group_info_service.upsert(
-            authority_provided_id=self._context.h_authority_provided_id,
-            consumer_key=self._request.lti_user.oauth_consumer_key,
-            params=self._request.params,
-        )
+        for group in groups:
+            self.h_api.add_user_to_group(h_user=h_user, group_id=group.groupid)
 
-    def _upsert_h_group(self, group_id, group_name, creator):
+    def _upsert_h_group(self, group, creator):
         """Update the group and create it if the user is an instructor."""
 
         try:
-            self.h_api.update_group(group_id=group_id, group_name=group_name)
+            self.h_api.update_group(group_id=group.groupid, group_name=group.name)
```

Comment:
We could push this value object down into h_api, but soon enough all these methods will be gone, so why bother?